### PR TITLE
CTPPS: Geometry for UFSDetectors

### PR DESCRIPTION
**Pull request by forthommel**
This PR defines the geometry for the UFSD planes replacing the diamonds' planes 4 in sectors 45 and 56 in 2017.
It also redefines a few diamond volumes' parentage to ease the alignment procedure for all subdetectors.

The re-reco of 2017 data with UFSDs is dependent on this PR.

`runTheMatrix.py -l limited` yielded `19 18 17 14 8 1 1 1 tests passed, 1 1 0 0 0 0 0 0 failed` (usual `4.22` DAS problem, and "Connection timed out problem" while opening `RecoPixelVertexing/PixelLowPtUtilities/data/pixelShapePhase1_noL1.par` in step2 of `7.3`)

### CODE REVIEW

**Comment by forthommel** [comment]
Updated the PR description with the result of `runTheMatrix.py` on top of d8323448c833404ce9855eb12696171df2d44e25.

**Comment by ianna** [comment]
Please test

File: Geometry/VeryForwardData/data/CTPPS_Diamond_Planes/CTPPS_Diamond_Plane1.xml [hunk 1]
```diff
@@ -29,12 +29,12 @@
 
   <PosPartSection label="CTPPS_Diamond_Plane1.xml">
 
-    <PosPart copyNumber="1">
-      <rParent name="CTPPS_Diamond_Plane"/>
-      <rChild name="CTPPS_Diamond_Detector_Assembly:PCB"/>
-      <Translation x="-[CTPPS_Diamond_Parameters:Diamond_dx]*2+[CTPPS_Diamond_Parameters:PCB_dx]/2+[CTPPS_Diamond_Parameters:PCB_extend_dx]" y="0*mm" z="([CTPPS_Diamond_Parameters:Diamond_dz]+[CTPPS_Diamond_Parameters:Metalized_thick]+[CTPPS_Diamond_Parameters:PCB_thick])/2"/>
-      <rRotation name="rotations:000D"/>
-    </PosPart>
+    <!--<PosPart copyNumber="1">-->
```
Comment (ianna) [inline Geometry/VeryForwardData/data/CTPPS_Diamond_Planes/CTPPS_Diamond_Plane1.xml]: @forthommel - please remove commented out lines

File: Geometry/VeryForwardData/data/CTPPS_Diamond_Detector_Assembly.xml [hunk 2]
```diff
@@ -61,12 +61,24 @@
       <Translation y="0*cm" x="[Shift_dx]" z="([CTPPS_Diamond_Parameters:Plane_dz]+[CTPPS_Diamond_Parameters:Plane_gap])/2"/>
       <rRotation name="CTPPS_Diamond_Transformations:planes_x_rotation"/>
     </PosPart>
+    <!--<PosPart copyNumber="4">-->
```
Comment (ianna) [inline Geometry/VeryForwardData/data/CTPPS_Diamond_Detector_Assembly.xml]: @forthommel - please remove commented out lines

File: Geometry/VeryForwardData/data/CTPPS_Diamond_Detector_Assembly.xml [hunk 3]
```diff
@@ -61,12 +61,24 @@
...
+      <!--<rParent name="CTPPS_Diamond_Main_Box"/>-->
+      <!--<rChild name="CTPPS_Diamond_Plane4:CTPPS_Diamond_Plane"/>-->
+      <!--<Translation y="0*cm" x="[Shift_dx]" z="([CTPPS_Diamond_Parameters:Plane_dz]+[CTPPS_Diamond_Parameters:Plane_gap])*3/2"/>-->
+      <!--<rRotation name="CTPPS_Diamond_Transformations:planes_x_rotation"/>-->
+    <!--</PosPart>-->
     <PosPart copyNumber="4">
       <rParent name="CTPPS_Diamond_Main_Box"/>
+      <rChild name="CTPPS_UFSD_Plane4:CTPPS_UFSD_Plane"/>
+      <Translation y="0*cm" x="[Shift_dx]" z="([CTPPS_Diamond_Parameters:Plane_dz]+[CTPPS_Diamond_Parameters:Plane_gap])*3/2"/>
+<!--       <rRotation name="CTPPS_Diamond_Transformations:planes_x_rotation"/> -->
```
Comment (ianna) [inline Geometry/VeryForwardData/data/CTPPS_Diamond_Detector_Assembly.xml]: @forthommel - please remove commented out lines

File: Geometry/VeryForwardData/data/CTPPS_Diamond_Detector_Assembly.xml [hunk 4]
```diff
@@ -61,12 +61,24 @@
...
+    </PosPart>
+<!--    <PosPart copyNumber="4">
```
Comment (ianna) [inline Geometry/VeryForwardData/data/CTPPS_Diamond_Detector_Assembly.xml]: @forthommel - please remove commented out lines

File: Geometry/VeryForwardData/data/CTPPS_Diamond_Planes/CTPPS_Diamond_Plane2.xml [hunk 5]
```diff
@@ -29,12 +29,12 @@
 
   <PosPartSection label="CTPPS_Diamond_Plane2.xml">
 
-    <PosPart copyNumber="2">
-      <rParent name="CTPPS_Diamond_Plane"/>
-      <rChild name="CTPPS_Diamond_Detector_Assembly:PCB"/>
-      <Translation x="-[CTPPS_Diamond_Parameters:Diamond_dx]*2+[CTPPS_Diamond_Parameters:PCB_dx]/2+[CTPPS_Diamond_Parameters:PCB_extend_dx]" y="0*mm" z="([CTPPS_Diamond_Parameters:Diamond_dz]+[CTPPS_Diamond_Parameters:Metalized_thick]+[CTPPS_Diamond_Parameters:PCB_thick])/2"/>
-      <rRotation name="rotations:000D"/>
-    </PosPart>
+    <!--<PosPart copyNumber="2">-->
```
Comment (ianna) [inline Geometry/VeryForwardData/data/CTPPS_Diamond_Planes/CTPPS_Diamond_Plane2.xml]: @forthommel - please remove commented out lines

File: Geometry/VeryForwardData/data/CTPPS_Diamond_Planes/CTPPS_Diamond_Plane3.xml [hunk 6]
```diff
@@ -29,12 +29,12 @@
 
   <PosPartSection label="CTPPS_Diamond_Plane3.xml">
 
-    <PosPart copyNumber="3">
-      <rParent name="CTPPS_Diamond_Plane"/>
-      <rChild name="CTPPS_Diamond_Detector_Assembly:PCB"/>
-      <Translation x="-[CTPPS_Diamond_Parameters:Diamond_dx]*2+[CTPPS_Diamond_Parameters:PCB_dx]/2+[CTPPS_Diamond_Parameters:PCB_extend_dx]" y="0*mm" z="([CTPPS_Diamond_Parameters:Diamond_dz]+[CTPPS_Diamond_Parameters:Metalized_thick]+[CTPPS_Diamond_Parameters:PCB_thick])/2"/>
-      <rRotation name="rotations:000D"/>
-    </PosPart>
+    <!--<PosPart copyNumber="3">-->
```
Comment (ianna) [inline Geometry/VeryForwardData/data/CTPPS_Diamond_Planes/CTPPS_Diamond_Plane3.xml]: @forthommel - please remove commented out lines

File: Geometry/VeryForwardData/data/CTPPS_Diamond_Planes/CTPPS_Diamond_Plane4.xml [hunk 7]
```diff
@@ -29,12 +29,12 @@
 
   <PosPartSection label="CTPPS_Diamond_Plane4.xml">
 
-    <PosPart copyNumber="4">
-      <rParent name="CTPPS_Diamond_Plane"/>
-      <rChild name="CTPPS_Diamond_Detector_Assembly:PCB"/>
-      <Translation x="-[CTPPS_Diamond_Parameters:Diamond_dx]*2+[CTPPS_Diamond_Parameters:PCB_dx]/2+[CTPPS_Diamond_Parameters:PCB_extend_dx]" y="0*mm" z="([CTPPS_Diamond_Parameters:Diamond_dz]+[CTPPS_Diamond_Parameters:Metalized_thick]+[CTPPS_Diamond_Parameters:PCB_thick])/2"/>
-      <rRotation name="rotations:000D"/>
-    </PosPart>
+    <!--<PosPart copyNumber="4">-->
```
Comment (ianna) [inline Geometry/VeryForwardData/data/CTPPS_Diamond_Planes/CTPPS_Diamond_Plane4.xml]: @forthommel - please remove commented out lines

File: Geometry/VeryForwardData/data/CTPPS_Diamond_Sensitive_Dets.xml [hunk 8]
```diff
@@ -0,0 +1,16 @@
+<?xml version="1.0"?>
+<DDDefinition xmlns="http://www.cern.ch/cms/DDL"
+              xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+              xsi:schemaLocation="http://www.cern.ch/cms/DDL ../../DDLSchema/DDLSchema.xsd">
+    <SpecParSection label="CTPPS_Diamond_Sensitive_Dets.xml">
+        <SpecPar name="ROUHitsCTPPSDiamond">
+            <!--  <PartSelector path="//Pat.*_Str.*:Diamond"/>-->
```
Comment (ianna) [inline Geometry/VeryForwardData/data/CTPPS_Diamond_Sensitive_Dets.xml]: @forthommel - please remove commented out lines

File: Geometry/VeryForwardData/data/CTPPS_Diamond_Sensitive_Dets.xml [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+
+            <PartSelector path="//CTPPS_Diamond_Pattern.*_Segment.*:CTPPS_Diamond_Segment"/>
+
+            <!--   <PartSelector path="//PPS_Timing_Box_.*:BoxOFlbar"/>-->
```
Comment (ianna) [inline Geometry/VeryForwardData/data/CTPPS_Diamond_Sensitive_Dets.xml]: @forthommel - please remove commented out lines

File: Geometry/VeryForwardData/data/CTPPS_Timing_Positive_Station.xml [hunk 10]
```diff
@@ -41,24 +41,24 @@
...
-      <Translation x="0*mm" y="0*mm" z="-([CTPPS_Timing_Station_Parameters:Station_Length]/2-[CTPPS_Timing_Station_Parameters:Station_Pipe_Length]/2)"/>
-    </PosPart>
-    <PosPart copyNumber="1">
-      <rParent name="CTPPS_Timing_Positive_Station:CTPPS_Timing_Positive_Station"/>
-      <rChild name="CTPPS_Timing_Positive_Station_Tube_2"/>
-      <rRotation name="rotations:000D"/>
-      <Translation x="0*mm" y="0*mm" z="([CTPPS_Timing_Station_Parameters:Station_Length]/2-[CTPPS_Timing_Station_Parameters:Station_Pipe_Length]/2)"/>
+      <Translation x="0*mm" y="0*mm" z="-([CTPPS_Timing_Station_Parameters:Station_Length]/2-([CTPPS_Timing_Station_Parameters:Station_Pipe_Length]+[RP_220_Right_Station:RP_220_Right_Prim_Hor_z]+[RP_Device:RP_Device_Length_z]/2)/2)"/>
     </PosPart>
+    <!--<PosPart copyNumber="1">-->
```
Comment (ianna) [inline Geometry/VeryForwardData/data/CTPPS_Timing_Positive_Station.xml]: @forthommel - please remove commented out lines

File: Geometry/VeryForwardData/data/CTPPS_UFSD_Planes/CTPPS_UFSD_Plane4.xml [hunk 11]
```diff
@@ -0,0 +1,272 @@
...
+  <LogicalPartSection label="CTPPS_UFSD_Plane4.xml">
+    <LogicalPart name="CTPPS_UFSD_Plane" category="unspecified">
+      <rSolid name="Plane"/>
+      <rMaterial name="materials:Air"/>
+    </LogicalPart>
+  </LogicalPartSection>
+
+  <PosPartSection label="CTPPS_UFSD_Plane4.xml">
+
+    <!--<PosPart copyNumber="4">-->
```
Comment (ianna) [inline Geometry/VeryForwardData/data/CTPPS_UFSD_Planes/CTPPS_UFSD_Plane4.xml]: @forthommel - please remove commented out lines

File: Geometry/VeryForwardData/data/RP_220_Left_Station.xml [hunk 12]
```diff
@@ -112,11 +112,11 @@
             <rChild name="RP_220_Left_Station_Tube_2"/>
             <Translation x="0*mm" y="0*mm" z="-(-[RP_220_Left_Station_Length]/2+([RP_220_Left_Prim_Vert_z]+[RP_220_Left_Prim_Hor_z])/2)"/>
         </PosPart>
-        <PosPart>
-            <rParent name="RP_220_Left_Station"/>
-            <rChild name="RP_220_Left_Station_Tube_3"/>
-            <Translation x="0*mm" y="0*mm" z="-(-[RP_220_Left_Station_Length]/2+([RP_220_Left_Prim_Hor_z]+[RP_220_Left_Sec_Hor_z])/2)"/>
-        </PosPart>
+        <!--<PosPart>-->
```
Comment (ianna) [inline Geometry/VeryForwardData/data/RP_220_Left_Station.xml]: @forthommel - please remove commented out lines

File: Geometry/VeryForwardData/data/RP_220_Right_Station.xml [hunk 13]
```diff
@@ -112,11 +112,11 @@
             <rChild name="RP_220_Right_Station_Tube_2"/>
             <Translation x="0*mm" y="0*mm" z="-[RP_220_Right_Station_Length]/2+([RP_220_Right_Prim_Vert_z]+[RP_220_Right_Prim_Hor_z])/2"/>
         </PosPart>
-        <PosPart>
-            <rParent name="RP_220_Right_Station"/>
-            <rChild name="RP_220_Right_Station_Tube_3"/>
-            <Translation x="0*mm" y="0*mm" z="-[RP_220_Right_Station_Length]/2+([RP_220_Right_Prim_Hor_z]+[RP_220_Right_Sec_Hor_z])/2"/>
-        </PosPart>
+        <!--<PosPart>-->
```
Comment (ianna) [inline Geometry/VeryForwardData/data/RP_220_Right_Station.xml]: @forthommel - please remove commented out lines

**Comment by forthommel** [comment]
Hello @ianna,
Many thanks for your fast review! I removed this extra set of comments in https://github.com/cms-sw/cmssw/pull/20548/commits/3823603a7cc4fe9c73876307debdfb3b9987a003. In the meantime, we discovered a slight offset in the UFSD plane definition (and simplified the volumes parentage) only affecting our local simulation part (hence, not the reconstruction itself). We introduced this fix in https://github.com/cms-sw/cmssw/pull/20548/commits/81fbf0753d7d287e372ae6c2fe4c10f5ba8269d5.

**Comment by ianna** [comment]
please test

**Comment by ianna** [comment]
+1

**Comment by davidlange6** [comment]
+1

**Pull request merged**